Feature/hdf 86 create getter setter

ernestpascual opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

## Description of Changes

- getter and setter for `minterContractAddress`
- getter and setter for `minterLpRewardsRatio`
- merged from lucas' branch
- revised calculation for getting HALOHALO APY
- current HALOHALO Price calculator
- new tests for HALOHALO and HALO changes
- revised breaking tests
- cleaned up tests


Notes for Rewards contract:
- @bitcoinbrisbane  you can check the test implementation of checking `pool.accHaloPerShare` in line 368. I stored first the initial `pool.accHaloPerShare` in line 340. My reason for this is since we are updating `pool.accHaloPerShare` again after calling `rewardsContract.minterLpPools` in line 343, the `pool.accHaloPerShare` value will either increase or remain the same.


Notes for halohalo contract:
- I removed the dependency of `rewardsContract.releaseVestedRewards()` and simulate this by minting more HALO inside the halohalo contract. 
- For calculating current value of HALOHALO in terms of HALO, I simulated the release of vested rewards by minting more halo. I then compared if the price increased from 1 in the previous check.
- For`HALO earned by User A > HALO earned by User B > HALO earned by User C`, I changed the release of vested rewards by minting more HALO to the rewards contract. I also added an `expect` to check the values of the 3 user addresses

Let me know if this works!

[Link to Jira Ticket](https://halodao.atlassian.net/browse/HDF-86)

## How To Test
- open up a node by entering `yarn run node`
- run the tests using `yarn run test`

### Developer Checklist:

- [x] I have followed the guidelines in our Contributing document
- [x] This PR has a corresponding JIRA ticket
- [x] My branch conforms with our naming convention i.e. `feature/HDF-XXX-description`
- [x] I have written new tests for your core changes, as applicable
- [x] I have successfully ran tests locally
- [x] I have formatted my code using format document in VSCode

### Reviewers Checklist:

- [ ] Code is readable and understandable; any unclear parts have explanations
- [ ] UI/UX changes match the corresponding figma/other design resources, if applicable
- [ ] I have successfully ran tests locally

ctverceles commented on test/halohalo.ts:
```diff
@@ -224,115 +54,70 @@ describe('HALOHALO Contract', async () => {
+        Number(await halohaloContract.genesisTimestamp()),
+        'Genesis timestamp is zero after entering HALO.'
+      ).to.be.greaterThan(0)
     })
```
@ernestpascual we can also add another test here that expects the `owner` address to have a certain HALOHALO balance (equal to the HALO tokens deposited in HaloHalo due to first time `enter` is being called but good to test this as well still)

ctverceles commented on test/halohalo.ts:
```diff
@@ -224,115 +54,70 @@ describe('HALOHALO Contract', async () => {
+      halohaloPrice = +formatEther(
+        await halohaloContract.getCurrentHaloHaloPrice()
       )
+      expect(halohaloPrice, 'HALOHALO is not equal to 1').to.equal(1)
```
@ernestpascual is `1` supposed to be the value of `ownerHaloBal` in line 70? Moving forward let's try to avoid having numbers in the expected price and instead store it in a variable named `expectedHALOHALOPrice` or something. 

Likewise, can we rename `halohaloPrice` to `actualHALOHALOPrice`

Then can also do this `expectedHALOPrice = ownerBal` after line 70, then at this line 88 we do `expect(actualHALOHALOPrice, 'HALOHALO is not equal to 1').to.equal(expectedHALOHALOPrice)`

ctverceles commented on test/halohalo.ts:
```diff
@@ -224,115 +54,70 @@ describe('HALOHALO Contract', async () => {
-      // TODO: Check value vested
+    it('Calculates current value of HALOHALO in terms of HALO after vesting', async () => {
+      // Simulate sending 20% of released rewards
+      await haloTokenContract.mint(halohaloContract.address, parseEther('120'))
```
@ernestpascual can we also be very verbose about where `120` came from? Either through comments or calculating it within this code. I'm not sure how this is calculated from reading this :)

ctverceles commented on test/halohalo.ts:
```diff
@@ -224,115 +54,70 @@ describe('HALOHALO Contract', async () => {
+        await halohaloContract.getCurrentHaloHaloPrice()
+      )
+      // Expect price to be greater than 1
+      expect(
```
@ernestpascual is it possible to expect `halohaloPriceAfterVesting` to be compared to an actual value rather than expecting to be greater than `halohaloPrice`? Just to be air tight with our tests :)

mystbrent replied:
Hi @ernestpascual thank you for doing the getCurrentHaloHaloPrice(). Can I request a more real life example of this function in the test cases? I think something similar to @ctverceles Alice, Bob example would be really great to illustrate how the logic here works. Thank you :)

ernestpascual replied:
@ctverceles I added an extra test instead. I believe we still need the first test to show that it's not 1:1.

ernestpascual replied:
@mystbrent  sure!

Alice wants to check the current price of her HALOHALO tokens in terms of HALO. The HALOHALO price is determined by the ratio of the total HALO balance of the HALOHALO contract and the total supply of HALOHALO minted by the contract. For Alice to get the price of her tokens, we multiply her number of tokens by the ratio of total HALO inside the contract and total supply of HALOHALO minted. 

I hope this answers your question 😅

bitcoinbrisbane replied:
In "Given, When, Then" https://martinfowler.com/bliki/GivenWhenThen.html

## Scenario 1
- Given Alice has 100 HALO
- And the HALO contract has 100 HALO
- When she deposits 100 HALO 
- Then she receives 100 HALOHALO

ctverceles commented on test/halohalo.ts:
```diff
@@ -224,115 +54,70 @@ describe('HALOHALO Contract', async () => {
 
-      const ownerXHalo = await halohaloContract.balanceOf(owner.address)
-      await halohaloContract.leave(ownerXHalo)
+      const ownerHaloHalo = await halohaloContract.balanceOf(owner.address)
```
Thanks for updating the vars @ernestpascual 🙏

ctverceles commented on test/halohalo.ts:
```diff
@@ -353,20 +138,17 @@ describe('HALOHALO Contract', async () => {
-      await rewardsContract.releaseVestedRewards()
+
+      // Simulated vesting by minting additional HALO to halohalo contract
+      await haloTokenContract.mint(halohaloContract.address, parseEther('120'))
```
@ernestpascual same comment about the 120 up there :) can we also be very verbose about where 120 came from? Either through comments or calculating it within this code. I'm not sure how this is calculated from reading this :)

ctverceles reviewed (changes requested):
@ernestpascual great job altering the tests and thanks for finishing up that function for @mystbrent to continue with his work. Logic looks great and tests all run on my local. Just some minor comments

What are your thoughts on implementing a best practice wherein whenever we test we adhere to the ff, which makes it easier for anyone in the team to understand what is happening (now that we're more devs):

- no "unlabeled numbers" in the values to be compared, always store a number in an `actual...` or `expected...` variable
- no "out of thin air" numbers in tests, always explain via comments or in code calculation where something came from

Following those practices, a sample test flow could be:

1. calculate some actual value, assign to `actual<variable_name>` (like `actualHALOHALOPrice`)
2. calculated expected value, assign to `expected<variable_name>` (like `expectedHALOHALOPrice`
3. compare actual vs expected value with clear err message (as you've done very well @ernestpascual )

Is this something you'd like to follow or is it overkill? FYI @bitcoinbrisbane if you have thoughts here

ctverceles commented on contracts/Rewards.sol:
```diff
@@ -33,935 +33,955 @@ import "hardhat/console.sol";
+    haloTokenAddress = _haloTokenAddress;
+    startingRewards = _startingRewards;
+    epochLength = _epochLength;
+    // minterLpRewardsRatio = _minterLpRewardsRatio;
```
@ernestpascual do we need these comments still?

ctverceles commented on contracts/Rewards.sol:
```diff
@@ -33,935 +33,955 @@ import "hardhat/console.sol";
+    return ammLpPoolsAddresses;
+  }
+
+  function getMinterContractAddress() public view returns (address) {
```
maybe we can also group the view functions under the VIEW FUNCTIONS comment block and check that the other comment blocks are still organising the rest of these funcs

bitcoinbrisbane commented on contracts/Rewards.sol:
```diff
@@ -33,935 +33,955 @@ import "hardhat/console.sol";
+  /// @notice utility constant
+  uint256 public constant BPS = 10**4;
+  uint256 public constant REWARD_PER_BLOCK = 29;
+  address private constant NULL_ADDRESS =
```
What about 
```suggestion
  address private constant NULL_ADDRESS = address(0);
```

ernestpascual replied:
thanks for this @bitcoinbrisbane !

bitcoinbrisbane commented on contracts/Rewards.sol:
```diff
@@ -33,935 +33,955 @@ import "hardhat/console.sol";
+    // minterLpRewardsRatio = _minterLpRewardsRatio;
+    ammLpRewardsRatio = _ammLpRewardsRatio;
+    vestingRewardsRatio = _vestingRewardsRatio;
+    //minterContract = _minter;
```
Lets remove
```suggestion
    //minterContract = _minter;
```

bitcoinbrisbane commented on contracts/Rewards.sol:
```diff
@@ -33,935 +33,955 @@ import "hardhat/console.sol";
+    uint256 amount
+  );
+  event MinterRewardPoolRatioUpdatedEvent(
+    address collateralAddress,
```
Maybe 
```suggestion
    address indexed collateralAddress,
```

bitcoinbrisbane commented on contracts/Rewards.sol:
```diff
@@ -33,935 +33,955 @@ import "hardhat/console.sol";
+    uint256 lastRewardBlock
+  );
+  event AmmLPRewardUpdatedEvent(
+    address lpAddress,
```
Maybe 

```suggestion
    address indexed lpAddress,
```

bitcoinbrisbane commented on contracts/Rewards.sol:
```diff
@@ -33,935 +33,955 @@ import "hardhat/console.sol";
+
+    pool.accHaloPerShare = pool.accHaloPerShare.add(
+      haloReward.mul(DECIMALS).div(minterCollateralSupply)
+      //haloReward
```
```suggestion
```

bitcoinbrisbane reviewed (changes requested):
Few suggestions

mystbrent reviewed (changes requested):
All tests are running great! Just requested an additional real life example to the HaloHalo.getCurrentHaloHaloPrice() in the test cases.

ernestpascual commented:
Updated with all the recommended changes!  ready to review again.
- broke down the old `onlyMinter()` modifier to` onlyMinter()` and `requireMinter()`
- added `requireMinter()` on functions that can be called by other parties besides the minter but need the `minterContractAddress`
- Removed the commented constructor parameters
- Delete unused comments
- deleted `@notice` of other view functions that are redundant. Left `@dev instead` 
- set `NULL_ADDRESS` to `address(0)` from @bitcoinbrisbane 's suggestion
- added indexed to address parameters of events
- changed all values as variables for the tests
- added additional testing as recommended by @ctverceles

bitcoinbrisbane reviewed (approved):
LGTM

ctverceles reviewed (approved):
Great work @ernestpascual ! all running on my local and changes look addressed
